downloader/src/main/kotlin/vcs/Git.kt

```diff
@@ -167,11 +167,21 @@ class Git : VersionControlSystem(), CommandLineTool {
         return getWorkingTree(targetDir)
     }
 
-    override fun updateWorkingTree(workingTree: WorkingTree, revision: String, path: String, recursive: Boolean) =
-        updateWorkingTreeWithoutSubmodules(workingTree, revision) && (!recursive || updateSubmodules(workingTree))
+    override fun updateWorkingTree(
+        workingTree: WorkingTree,
+        revision: String,
+        path: String,
+        recursive: Boolean
+    ): Result<String> =
+        updateWorkingTreeWithoutSubmodules(workingTree, revision).mapCatching {
+            // In case this throws the exception gets encapsulated as a failure.
+            if (recursive) updateSubmodules(workingTree)
+
+            revision
+        }
 
-    private fun updateWorkingTreeWithoutSubmodules(workingTree: WorkingTree, revision: String): Boolean {
-        try {
+    private fun updateWorkingTreeWithoutSubmodules(workingTree: WorkingTree, revision: String): Result<String> {
+        runCatching {
```

Comment:
Not sure how this would look like in practice, but maybe the multiple `runCatching() { ... }.onSuccess { ... }`  constructs can be replaced by a chain of `.map()` and `.recover()` calls?

Reply:
Yes, that again looks more stylish 😁 Done. (But please note this PR is still a draft.)